# Outline:

## Introduction: 
### 1. Phylogenetics is central to combatting the spread of novel and endemic pathogens. 
+ Pathogen genome sequencing has made phylogenetics central to combatting the pandemic. Examples. 
+ The vast datasets have also shattered typical phylogenetics workflows.
+ The is a need to evaluate new (or in this case old!) approaches for maintaining a real-time up to date comprehensive phylogeny. 

### 2. Imporant differences relative to previous pathogen tracing efforts are caused by the dramatically different pandemic sampling scheme.  
+ During the SARS-CoV-2 pandemic, the vast scale of viral genome sequencing means that samples are often extremely genetically similar or identical. 
+ This is likely to be true in the future for all pathogen genomics efforts going forward. 
+ "State of the art" maximum likelihood approaches are expected to outperform parsimony methods when branch lengths are long. 
+ However, when branches are short, only rarely will two mutations occur on the same branch and the advantages of ML approaches is not clear.  

### 3. Data flows in constantly, this presents an additional major challenge for typical phylgoenetics workflows. 
+ In a typical, *de novo*, phylogenetics workflow, the tree is reinferred starting from scratch periodically. 
+ During a pandemic, data stream in constantly, requiring *de novo* phylogenetics approaches to reinfer the tree very frequently. 
+ As the existing datasets grow, * adding new genome sequenes to the ever-growing phylogeny becomes appealing. 
+ Important, heretofore unresolved, issues surround how to most accurately and efficiently generate and optimize these vast phylogenies. 

### 4. We propose and evaluate a suite of approaches for maintaining a comprehensive globaly phylogeny comprised of XXX SARS-CoV-2 genomes. 
+ Summary of main points. 


## Results and Discussion: 
### 1. 
